firebase_admin/db.py

```diff
@@ -0,0 +1,641 @@
+        return Query(order_by=path, client=self._client, pathurl=self._add_suffix())
+
+    def order_by_key(self):
+        """Creates a Query that orderes data by key.
+
+        Returned Query can be used to set additional parameters, and execute complex database
+        queries (e.g. limit queries, range queries).
+
+        Returns:
+          Query: A database Query instance.
+        """
+        return Query(order_by='$key', client=self._client, pathurl=self._add_suffix())
+
+    def order_by_value(self):
+        """Creates a Query that orderes data by value.
+
+        Returned Query can be used to set additional parameters, and execute complex database
+        queries (e.g. limit queries, range queries).
+
+        Returns:
+          Query: A database Query instance.
+        """
+        return Query(order_by='$value', client=self._client, pathurl=self._add_suffix())
+
+    def order_by_priority(self):
+        """Creates a Query that orderes data by priority.
+
+        Returned Query can be used to set additional parameters, and execute complex database
+        queries (e.g. limit queries, range queries). Due to a limitation of the
+        underlying REST API, the order-by-priority constraint can only be enforced during
+        the execution time of the Query. When the Query returns results, the actual results
+        will be returned as an unordered collection.
+
+        Returns:
+          Query: A database Query instance.
+        """
+        return Query(order_by='$priority', client=self._client, pathurl=self._add_suffix())
+
+    def _add_suffix(self, suffix='.json'):
+        return self._pathurl + suffix
+
+    @classmethod
+    def _check_priority(cls, priority):
+        if isinstance(priority, six.string_types) and priority.isalnum():
+            return
+        if isinstance(priority, numbers.Number):
+            return
+        raise ValueError('Illegal priority value: "{0}". Priority values must be numeric or '
+                         'alphanumeric.'.format(priority))
+
+
+class Query(object):
+    """Represents a complex query that can be executed on a Reference.
+
+    Complex queries can consist of up to 2 components - a required ordering constraint, and an
+    optional filtering constraint. At the server, data is first sorted according to the given
+    ordering constraint (e.g. order by child). Then the filtering constraint (e.g. limit, range)
+    is applied on the sorted data to produce the final result. Despite the ordering constraint,
+    the final result is returned by the server as an unordered collection. Therefore the Query
+    interface performs another round of sorting at the client-side before returning the results
+    to the caller. This client-side sorted results are returned to the user as a Python
+    OrderedDict. However, client-side sorting is not feasible for order-by-priority queries.
+    Therefore for such queries results are returned as a regular unordered dict.
```

Comment:
Unfortunately not. This is a limitation of the REST API. It doesn't report priorities of individual elements when returning results, and therefore at the client-side we don't have anything to sort by. I've been in contact with the database server folks, and they confirmed this is going to require a change at their end to implement. In the meantime I've added a note to the API proposal raising this issue. One option we are considering is dropping support for order-by-priority queries, at least until the server-side support is available for that.